### ISSUE TITLE
Teaching best-practices by awarding bonus brownie points for optional tests.

### ISSUE THREAD

**Issue by davinderpalrehal** [freeCodeCamp/freeCodeCamp#16463]
#16458 
Followup on the issue.

Campers will be encouraged to research on best practices that pertain to certain tags in HTML and other languages. For example adding the "for" attribute to labels, checkboxes and other elements. Following best practices would award the campers brownie points which could be added to their public profile.

**Comment by raisedadead** [comment]
/cc @freeCodeCamp/moderators 

I think this is a cool idea, this way we can recommend best-practices and at the same time, keep some optional tests which may be done for the incentive of more points.

Thoughts?

**Comment by krisb1220** [comment]
I like this!!! Seems a bit heavy though :P Contributors?  :P

**Comment by raisedadead** [comment]
Yes, even if this approved, we can take this only after beta is shipped.

**Comment by erictleung** [comment]
I second this feature if it could be implemented! This should address any opinionated styling and best practices discussions that our curriculum may have :+1: But I also really like this feature conceptually to reward those to do more, especially when they are best practices :smile: 

This feature will likely be on the back burner for a while as @raisedadead mentioned because of beta. For long-term thinking, we might also want to consider:

- How to determine what should be considered best practices for each challenge?
- Should they be available for all interactive challenges?
- How many extra opportunities should be available?
- Also, could this idea be extended to algorithms as well to incentivize campers to make a better and more robust function/algorithm?

**Comment by davinderpalrehal** [comment]
I would like to work on this whenever it is taken off the backburner. I will need help with determining the best practices just like @erictleung mentioned.

**Comment by Greenheart** [comment]
@davinderpalrehal This is a great idea! Thanks for starting the discussion :smile:

Along with the questions earlier in this thread, the question that remains is how we prioritize this among other things we want to do. Another thought is to start small and try to limit the scope of the initial version of such a feature, to actually get something launched :)

**Comment by systimotic** [comment]
I agree with all the thoughts above. I think we can do a lot better in teaching best practices.

I do want to note that we should be careful that we don't make existing challenges more complicated than they need to be by doing this. Learning to code is already hard enough without having best practices thrown around from the beginning 😅 

Dedicated challenges to best practices might be helpful too.

**Comment by raisedadead** [comment]
Yes, totally agree with you, the intent should always be:

- To be succinct instructions and not overwhelming the users, with a overload of information.
- A a few best-practice norms in some challenges, (NOT every challenge).

And keep a balance, by not trying to over complicate the verbiage in anyway.

**Comment by raisedadead** [comment]
@QuincyLarson 

Yes, you are right https://github.com/freeCodeCamp/freeCodeCamp/issues/6305 is related. 
Well, its a very old issue on the roadmap. 

But, this would be a really interesting feature, like you said 1 point for a project seems less, adding the granularity will be good thing.

**Do you think we can call them something other than "Brownie Points" ?**

This is because, some non native campers often get confused, with the no thinking that they are challenge no.s, the issue that I linked above, wants to normalize that so the increments are done in 5, 10, etc, and not 1 point, per challenge.

**Comment by raisedadead** [comment]
P.S I found your comment from another issue https://github.com/freeCodeCamp/freeCodeCamp/issues/12460#issuecomment-310221195 where you say, that they could be called XP. Which seems really logical to me.

**Comment by QuincyLarson** [comment]
@raisedadead Yes - I think XP is something pretty universal. Brownie Points are colloquial here in the US, and related to boy scouts and girl scouts - thus camp-themed. But XP is simpler and more easily understood. 

Changing the name won't fix the confusion around that number, though. What we need to do is multiply the number times 100. That way campers definitely won't confuse the number for the current challenge number. And in most video games, score is a multiple of 100.

**Comment by Ethan-Arrowood** [comment]
Now I am imagining a future FreeCodeCamp where users have customizable avatars and campers can redeem their XP for special items and stuff 😂 

In all seriousness though I really like this idea. What pops into my head as a possible technical implementation is an independent API from the main FreeCodeCamp curriculum page. This way we can implement this XP system across all FreeCodeCamp things. I'm already getting ideas for rewarding (meaningful) contributions (like if someones PR is merged they earn XP), and this is all possible by integrating the GitHub API and hopefully this FreeCodeCamp XP API

**Comment by raisedadead** [comment]
Yes, Ethan. We have an Open API project on the roadmap just for such things.

**Comment by raisedadead** [comment]
Yes, Ethan. We have an Open API project on the roadmap just for such things.

**Comment by QuincyLarson** [comment]
@Ethan-Arrowood Yeah - it would be cool to be able to redeem these points for virtual goods, or have achievements. This was something we discussed about two years ago. I imagine such an ecosystem could crop up on external websites as well, through consuming the open API.

**Comment by QuincyLarson** [comment]
Per discussion on https://github.com/freeCodeCamp/freeCodeCamp/issues/16458 we've decided to not pursue optional tests for the time being. I've elaborated on the reasoning behind this decision here: https://github.com/freeCodeCamp/freeCodeCamp/issues/16458#issuecomment-363312711
